ibis/common/deferred.py

```diff
@@ -0,0 +1,608 @@
+from __future__ import annotations
+
+import collections.abc
+import functools
+import inspect
+import operator
+from abc import abstractmethod
+from typing import Any, Callable, TypeVar, overload
+
+from ibis.common.bases import Final, FrozenSlotted, Hashable, Immutable, Slotted
+from ibis.common.collections import FrozenDict
+from ibis.common.typing import Coercible, CoercionError
+from ibis.util import Namespace, PseudoHashable, is_iterable
+
+
+class Deferred(Coercible, Hashable):
+    """Specification about constructing a value given a context.
+
+    The context is a dictionary that contains all the captured values and
+    information relevant for the builder.
+
+    The builder is used in the right hand side of the replace pattern:
+    `Replace(pattern, builder)`. Semantically when a match occurs for the
+    replace pattern, the builder is called with the context and the result
+    of the builder is used as the replacement value.
+    """
+
+    @classmethod
+    def namespace(cls, module):
+        return Namespace(DeferredProxy, module)
+
+    @abstractmethod
+    def resolve(self, context: dict):
+        """Construct a new object from the context.
+
+        Parameters
+        ----------
+        context
+            A dictionary containing all the captured values and information
+            relevant for the deferred.
+
+        Returns
+        -------
+        The constructed object.
+        """
+
+    @abstractmethod
+    def __eq__(self, other: Deferred) -> bool:
+        ...
+
+    @classmethod
+    def __coerce__(cls, value):
+        if isinstance(value, cls):
+            return value
+        elif isinstance(value, DeferredProxy):
+            return value._deferred
+        else:
+            raise CoercionError(f"Cannot coerce {value!r} to {cls.__name__!r}")
+
+
+class DeferredProxy(Slotted, Immutable, Final):
```

Comment:
When we present `_` to users, we talk about deferred objects and deferred expressions. I personally think it'd be easier to explain and document if instances of those expressions were called `Deferred` rather than `DeferredProxy`. Use the simpler name for the user-facing thing.

Perhaps call this `Deferred` and the wrapped instances `DeferredOp` or `DeferredExpr`? :shrug:.

Reply:
Ideally it should be transparent to the users, at least we shouldn't expose neither of the types for public consumption, so how about deferring the decision? :)

I'm actually kinda satisfied with `DeferredProxy` since it does nothing but proxying to provide syntactic sugar.

Reply:
> so how about deferring the decision?

I mean, if we're deferring the decision, then we should keep the class name the same as it was before this PR - `Deferred` :).

I wouldn't risk a naming bikeshed discussion if I didn't feel strongly about this. We need to call these things something in the documentation - currently we call them "deferred expressions" not "deferred proxies". These types will also show up in user's interactive sessions when introspecting objects, as well as in type annotations. Since the wrapped objects *are not user facing* then the naming of them feels less important to me. I don't see a reason to reserve the shorter/better `Deferred` name for an internal implementation detail forcing the user of `DeferredProxy` for the public type.

Further, there are cases in downstream APIs where we do need to access the type `Deferred` directly for introspection reasons (we do this in `ibisml` for example). Changing the class name here feels like a needless breaking change.

Reply:
@kszucs What's the resolution here? Seems like we should keep `Deferred` for now to avoid breaking ibisml

Reply:
I renamed the `Builder => Resolver` and kept `Deferred` as it used to be, no breakage.